### PR TITLE
feat(Core/SAI): Implement New Smart Actions

### PR DESCRIPTION
**Pull request by Gultask**
### SET_SCALE

- For custom use when spells aren't enough or serverside spells aren't found
- param1 is the integer part, param2 is the decimal part. So param1 = 5, param2 = 25 is like setting the scale to `5.25`

### SUMMON_RADIAL

- Summons a number of creatures within specified parameters, see video

### SET_ORIENTATION adjustments

- Allows you to set an angle from which the creature will turn instead of using only absolute orientation values

## Issues Addressed:

- Closes nothing

## Tests Performed:

- Tested ingame. I hoped Gobs would work as well but they don't 😢

## How to Test the Changes:

For SET_SCALE:
```
DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` = 448);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(448, 0, 0, 0, 38, 0, 100, 0, 1, 1, 0, 0, 0, 227, 0, 50, 0, 0, 0, 0, 9, 0, 0, 200, 0, 0, 0, 0, 0, ''),
(448, 0, 1, 0, 38, 0, 100, 0, 1, 2, 0, 0, 0, 227, 1, 0, 0, 0, 0, 0, 9, 0, 0, 200, 0, 0, 0, 0, 0, ''),
(448, 0, 2, 0, 38, 0, 100, 0, 1, 3, 0, 0, 0, 227, 1, 50, 0, 0, 0, 0, 9, 0, 0, 200, 0, 0, 0, 0, 0, ''),
(448, 0, 3, 0, 38, 0, 100, 0, 1, 4, 0, 0, 0, 227, 2, 0, 0, 0, 0, 0, 9, 0, 0, 200, 0, 0, 0, 0, 0, ''),
(448, 0, 4, 0, 38, 0, 100, 0, 1, 5, 0, 0, 0, 227, 2, 50, 0, 0, 0, 0, 9, 0, 0, 200, 0, 0, 0, 0, 0, ''),
(448, 0, 5, 0, 38, 0, 100, 0, 1, 6, 0, 0, 0, 227, 3, 0, 0, 0, 0, 0, 9, 0, 0, 200, 0, 0, 0, 0, 0, ''),
(448, 0, 6, 0, 38, 0, 100, 0, 1, 7, 0, 0, 0, 227, 0, 10, 0, 0, 0, 0, 9, 0, 0, 200, 0, 0, 0, 0, 0, '');
```
`.go c id 448`
`.npc set data 1 1`
`.npc set data 1 2`
`.npc set data 1 3`
`.npc set data 1 ...`

For SUMMON_RADIAL and SET_ORIENTATION:
```
DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` = 448);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(448, 0, 0, 0, 38, 0, 100, 0, 1, 1, 0, 0, 0, 228, 448, 5000, 4, 0, 90, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Data Set 1 1 - Summon Radial (4 Directions)'),
(448, 0, 1, 0, 38, 0, 100, 0, 1, 2, 0, 0, 0, 228, 448, 5000, 4, 0, 30, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Data Set 1 1 - Summon Radial (4 Directions)'),
(448, 0, 2, 0, 38, 0, 100, 0, 1, 3, 0, 0, 0, 228, 448, 5000, 3, 0, 120, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Data Set 1 1 - Summon Radial (4 Directions)'),
(448, 0, 3, 0, 38, 0, 100, 0, 1, 4, 0, 0, 0, 228, 448, 5000, 18, 0, 10, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Data Set 1 1 - Summon Radial (4 Directions)'),
(448, 0, 4, 0, 54, 0, 100, 0, 0, 0, 0, 0, 0, 11, 69823, 2, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Just Summoned - Cast Typhoon'),
(448, 0, 5, 0, 38, 0, 100, 0, 2, 1, 0, 0, 0, 22, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - On Data Set 2 1 - Start Turning 30 Every 1200ms'),
(448, 0, 6, 0, 60, 1, 100, 0, 1200, 1200, 1200, 1200, 0, 66, 0, 0, 30, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Hogger - Every 1200ms - Turn 30');
```
`.go c id 448`
`.npc set data 2 1` - Start turning 30 degrees every 1200ms
`.npc set data 1 2` 
`.npc set data 1 3`
`.npc set data 1 4`

https://github.com/azerothcore/azerothcore-wotlk/assets/100873791/4dce47bf-0b83-4a17-8c40-1407ff53aeb3

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by elthehablo** [comment]
For what kind of things is this for?

**Comment by Gultask** [comment]
For now? Nothing in particular. I guess custom stuff or whenever the spell can't be found OR when it's too bothersome to make a brand new serverside one.

The only other workaronds were scripting in cpp or making a new entry in spell_dbc afaik

**Comment by AnchyDev** [comment]
I think you have to respawn the gob with a new GUID with the new scale if you want gobs to work for the client.
That's how I recall ArcEmu working in the past.

**Comment by Gultask** [comment]
> I think you have to respawn the gob with a new GUID with the new scale if you want gobs to work for the client. That's how I recall ArcEmu working in the past.

Yeah that's out of my skillset 😛

**Comment by AnchyDev** [comment]
> > I think you have to respawn the gob with a new GUID with the new scale if you want gobs to work for the client. That's how I recall ArcEmu working in the past.
> 
> Yeah that's out of my skillset 😛

It looks a bit funky anyway as you are effectively respawning it, so the gob fades out and back in for a second.

**Comment by Kitzunu** [comment]
> param1 is the integer part, param2 is the decimal part

Why not make it a float to use only 1 param?

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -2824,6 +2824,21 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
             }
             break;
         }
+        case SMART_ACTION_SET_SCALE:
+        {
+            uint32 scaleInteger = e.action.setScale.scaleInteger;
+            uint32 scaleFraction = e.action.setScale.scaleFraction;
+            float scale = float(scaleInteger) + float(scaleFraction) / std::pow(10, std::floor(std::log10(float(scaleFraction ? scaleFraction : 1)) + 1));
```
Comment (Kitzunu) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:2831]:
```suggestion
            float scale = float(scaleInteger) + (float(scaleFraction) / 100);
```

Wouldn't this do the same thing, but easier?

scaleInteger = 2
scaleFraction = 75

2 + (75 / 100) = 2 + 0.75

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 2]
```diff
@@ -694,8 +694,9 @@ enum SMART_ACTION
     SMART_ACTION_ATTACK_STOP                        = 224,    //
     SMART_ACTION_SET_GUID                           = 225,    // Sends the invoker's or the base object's own ObjectGuid to target
     SMART_ACTION_DISABLE                            = 226,    // Disable the targeted creatures, setting them Invisible and Immune to All
+    SMART_ACTION_SET_SCALE                          = 227,    // Modifies the target's scale
```
Comment (Kitzunu) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h:697]: personally I think the comment should be reserved to what the fields are. Like
`SMART_ACTION_RANDOM_MUSIC                       = 217,    // SoundId1, SoundId2, SoundId3, SoundId4, onlySelf, type`

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 1 above)
Comment (Gultask) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:2831]: idk I just copied from SET_MOVEMENT_SPEED which is action 136, iirc

Comment (Kitzunu) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:2831]: Ah, well I dont see a reason not to just divde by 100

Comment (Kitzunu) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:2831]: And I havent checked this, but do you have a check for scaleFraction being greater than 99?

Comment (Gultask) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:2840]: I will check today after I wake up

**Comment by Gultask** [comment]
Since I already had this PR open I decided to expand it a bit

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 1 above)
Comment (Gultask) [reply src/server/game/AI/SmartScripts/SmartScript.cpp]: Changed to a percentage-based system instead

**Comment by Gultask** [comment]
managed to add distance offset :)
image

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**